### DISCUSSION TITLE
Generic build of POCL for various architectures

### DISCUSSION THREAD

**Discussion by maleadt** [pocl/pocl#1520] [Q&A]
Hi,

I'm working on packaging POCL for use in Julia, with the goal of using the CPU back-end to execute GPU-like code. We generate binaries that we distribute to our users, so I need a portable build that contains a generic version of the kernellib. I want to target i686/x86_64/armv6/armv7/aarch64 on Linux/Darwin (and hopefully Windows at some point).

Looking at e.g. Arch's packaging scripts, I figured that I need `KERNELLIB_HOST_CPU_VARIANTS=distro` in order to generate several variants of the kernel library. However, that doesn't seem to work on all platforms: https://github.com/pocl/pocl/blob/e9e39f03af5e71e4e65e9f92de71db78e595a7a4/CMakeLists.txt#L1429-L1430

Debian seems to carry two additional patches that enable a `GENERIC` CPU variant:
- enabling generation of the variant: https://salsa.debian.org/opencl-team/pocl/-/blob/main/debian/patches/distro.patch
- adding support for the `GENERIC` CPU type: https://salsa.debian.org/opencl-team/pocl/-/blob/main/debian/patches/generic-cpu.patch

Is this a reasonable approach, should it be upstreamed, or is there a better way to enable a fully generic build of POCL for these architectures?

**Comment by pjaaskel** [comment]
Hi! That's a reasonable but not ideal approach performance wise. [The wiki](https://github.com/pocl/pocl/wiki/Install-time-or-run-time-built-kernel-builtin-libraries-OUTDATED) has some (outdated) discussion on the topic and the possible more portable solutions. @anbe42 do you have comments / new ideas for this question?

**Comment by maleadt** [comment]
I agree that performance probably won't be the best, but there's currently no way to build POCL generically for non-x86/ppc architectures, so any improvement seems like a good first step to me.

In fact, I tried a simpler approach, where the `distro` build mode unconditionally adds a `generic` kernellib build and uses that as the `kernellib_fallback` option (which AFAICT seems unused right now?): https://github.com/maleadt/pocl/commit/e66a1726069764b0ee2b022206ab4356450b2eba
Would that be a reasonable thing to include, awaiting a better solution as discussed in the wiki page you linked?

**Reply by pjaaskel** [reply]
Sorry, missed this. Yep, let's pull that in.
